applicationset/controllers/applicationset_controller.go

```diff
@@ -1312,4 +1316,69 @@ func syncApplication(application argov1alpha1.Application, prune bool) (argov1al
 	return application, nil
 }
 
+var ownsHandler = predicate.Funcs{
+	CreateFunc: func(e event.CreateEvent) bool {
+		// if we are the owner and there is a create event, we most likely created it and do not need to
+		// re-reconcile
+		log.Debugln("received create event from owning an application")
+		return false
+	},
+	DeleteFunc: func(e event.DeleteEvent) bool {
+		log.Debugln("received delete event from owning an application")
+		return true
+	},
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		log.Debugln("received update event from owning an application")
+		appOld, isApp := e.ObjectOld.(*argov1alpha1.Application)
+		if !isApp {
+			return false
+		}
+		appNew, isApp := e.ObjectNew.(*argov1alpha1.Application)
+		if !isApp {
+			return false
+		}
+		requeue := shouldRequeueApplicationSet(appOld, appNew)
+		log.Debugf("requeue: %t caused by application %s\n", requeue, appNew.Name)
+		return requeue
+	},
+	GenericFunc: func(e event.GenericEvent) bool {
+		log.Debugln("received generic event from owning an application")
+		return true
+	},
+}
+
+// shouldRequeueApplicationSet determines when we want to requeue an ApplicationSet for reconciling based on an owned
+// application change
+// The applicationset controller owns a subset of the Application CR.
+// We do not need to re-reconcile if parts of the application change outside the applicationset's control.
+// An example being, Application.ApplicationStatus.ReconciledAt which gets updated by the application controller.
+// Additionally, Application.ObjectMeta.ResourceVersion and Application.ObjectMeta.Generation which are set by K8s.
+func shouldRequeueApplicationSet(appOld *argov1alpha1.Application, appNew *argov1alpha1.Application) bool {
+	if appOld == nil || appNew == nil {
+		return false
+	}
+
+	// the applicationset controller owns the application spec, labels, annotations, and finalizers on the applications
+	if !reflect.DeepEqual(appOld.Spec, appNew.Spec) ||
+		!reflect.DeepEqual(appOld.ObjectMeta.GetAnnotations(), appNew.ObjectMeta.GetAnnotations()) ||
+		!reflect.DeepEqual(appOld.ObjectMeta.GetLabels(), appNew.ObjectMeta.GetLabels()) ||
+		!reflect.DeepEqual(appOld.ObjectMeta.GetFinalizers(), appNew.ObjectMeta.GetFinalizers()) {
+		return true
+	}
+
+	// progressive syncs use the application status for updates. if they differ, requeue to trigger the next progression
+	if appOld.Status.Health.Status != appNew.Status.Health.Status || appOld.Status.Sync.Status != appNew.Status.Sync.Status {
+		return true
+	}
+
+	if appOld.Status.OperationState != nil && appNew.Status.OperationState != nil {
+		if appOld.Status.OperationState.Phase != appNew.Status.OperationState.Phase ||
+			appOld.Status.OperationState.StartedAt != appNew.Status.OperationState.StartedAt {
+			return true
+		}
+	}
```

Comment:
Should the requeue happen on this case, where OperationState is initially set from nothing?
```
appOld.Status.OperationState == nil && appNew.Status.OperationState != nil
```

I'm not actually sure if this field generally always exists for an Application.

Reply:
> Should the requeue happen on this case, where OperationState is initially set from nothing?
Maybe? It will happen for every application created when the application controller first syncs it. 

> I'm not actually sure if this field generally always exists for an Application.
Which field specifically? I see the [Application controller](https://github.com/argoproj/argo-cd/blob/b1376e9b5fd629d35df9fbf1f13ef6462c3d6da4/controller/appcontroller.go#L1124) setting the `OperationState` struct on the applications.

Reply:
`status.OperationState` specifically. I think it's fine to assume it will exist for requeue purposes.